File: quad/towr/towr_cd.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-
"""
TOWR-in-CasADi — 모델기반 지형 궤적최적화(오프라인).
TOWR(Winkler 2018) 알고리즘을 CasADi+IPOPT로 우리 스택(MuJoCo-native)에 재구현.
C++ TOWR(ifopt/ROS/catkin) 빌드 없이, casadi 3.7 번들 IPOPT 사용.

핵심(A/B/D1이 못하는 것):
  ▸ footholds(발 착지점)를 지형 위 결정변수로 최적화 → 갭/계단 회피·정밀 배치
  ▸ 스윙 발이 지형 위로 클리어런스
  ▸ SRBD(단일강체) 동역학 + 마찰콘 + ROM(도달성) 제약
  ▸ base 높이가 지형 따라 상승(TAMOLS 등가, 여기선 최적화가 자동 결정)
출력: base pos/vel/ori, 발 위치, 접촉력 궤적 → MuJoCo WBIC/PD 추종(Phase1).

기반: proxddp/simple_mpc env (casadi 3.7 + IPOPT + pinocchio).
실행:  /home/jsh/simple-mpc/.pixi/envs/default/bin/python towr_cd.py
"""
import numpy as np, casadi as ca, os, json

# ───────── 로봇 SRBD 파라미터 (pinocchio URDF 추출값, [[b-elevation-tamols-towr-track]]) ─────────
MASS = 38.016                                   # kg (총질량)
G    = 9.81
INER = np.array([0.941, 2.521, 2.236])          # centroidal 회전관성 diag Ixx,Iyy,Izz
BASE_H = 0.50                                    # 명목 base 높이
MU   = 0.6                                       # 마찰계수
# 공칭 발위치(base 로컬, 서있는 자세) — ROM 박스 중심
FEET = ['FL','FR','HL','HR']
P_NOM = {'FL':np.array([ 0.30, 0.16]), 'FR':np.array([ 0.30,-0.16]),
         'HL':np.array([-0.30, 0.16]), 'HR':np.array([-0.30,-0.16])}
ROM_DXY = np.array([0.13, 0.09])                # 발 xy가 공칭±ROM_DXY 내(도달성)
FOOT_Z_LO, FOOT_Z_HI = -0.56, -0.40            # 발 z(base 로컬) 범위(다리 길이)
F_MAX = 2.0*MASS*G                              # 접촉력 상한(수직)


def terrain_height(x, y, kind='flat', **kw):
    """지형 높이 h(x,y). CasADi 심볼릭·numpy 양쪽 동작(smooth). kind별."""
    if kind == 'flat':
        return 0.0*x
    if kind == 'step':                          # x=x0에서 h만큼 올라가는 계단(smooth)
        x0 = kw.get('x0', 1.0); h = kw.get('h', 0.10); s = kw.get('sharp', 40.0)
        return h*0.5*(1+ca.tanh(s*(x-x0))) if isinstance(x, (ca.SX,ca.MX,ca.DM)) \
               else h*0.5*(1+np.tanh(s*(x-x0)))
    if kind == 'gap':                           # [x0,x1] 구간이 깊게 파인 갭(smooth 우물)
        x0=kw.get('x0',0.9); x1=kw.get('x1',1.2); d=kw.get('depth',0.30); s=kw.get('sharp',60.0)
        th = ca.tanh if isinstance(x,(ca.SX,ca.MX,ca.DM)) else np.tanh
        return -d*0.5*(th(s*(x-x0)) - th(s*(x-x1)))
    if kind == 'platgap':                       # 높은 플랫폼(plat) + 갭 슬롯(floor 노출)
        x0=kw.get('x0',1.0); x1=kw.get('x1',1.3); plat=kw.get('plat',0.20); s=kw.get('sharp',60.0)
        th = ca.tanh if isinstance(x,(ca.SX,ca.MX,ca.DM)) else np.tanh
        return plat - plat*0.5*(th(s*(x-x0)) - th(s*(x-x1)))
    return 0.0*x


def build_and_solve(kind='flat', tkw=None, N=40, dt=0.02, Tg=0.40, duty=0.5,
                    x_goal=0.8, vx_des=None, verbose=True, gait='trot', phase_off=0.0):
    """
    phase-based TO. footholds=지형 위 변수.
      gait: 'trot'(대각쌍 FL·HR/FR·HL, 동적) | 'crawl'(한발씩 스윙, 3발지지 정적안정)
      N/dt: 노드·간격 → 총시간 T=N·dt ; Tg: gait 주기, duty: stance 비율
      x_goal: 전진 목표(m)
    반환: dict(궤적) 또는 None(실패).
    """
    tkw = tkw or {}
    T = N*dt
    if vx_des is None: vx_des = x_goal/T
    th = lambda x,y: terrain_height(x,y,kind,**tkw)
    # base 높이 참조: 갭 위에서 base가 갭 바닥으로 빠지면 안 됨 → 지지면(플랫폼) 레벨 사용.
    #   platgap=플랫폼 상수레벨, 그 외=지형 추종.
    _bl = tkw.get('plat', 0.0) if kind=='platgap' else None
    bhref = (lambda x,y: _bl) if _bl is not None else th

    # ── 접촉 스케줄: 각 발이 각 노드에서 stance인지(1/0) ──
    _CRAWL_ORDER = ['FL','HR','FR','HL']         # crawl 스윙 순서(지지삼각형 유지)
    def in_stance(foot, k):
        t = k*dt; ph = ((t + phase_off*Tg) % Tg)/Tg   # 0~1 (phase_off=전역 위상 오프셋)
        if gait == 'crawl':                       # 4구간, 각 구간서 1발만 스윙(swing=duty만큼)
            idx = _CRAWL_ORDER.index(foot)
            win_lo = idx*0.25; sw = 0.25*duty     # 이 발의 스윙 창
            return not (win_lo <= ph < win_lo+sw)  # 스윙 창 밖이면 stance
        off = 0.0 if foot in ('FL','HR') else 0.5  # trot: 반주기 오프셋
        return ((ph - off) % 1.0) < duty
    contact = {f:[in_stance(f,k) for k in range(N+1)] for f in FEET}

    opti = ca.Opti()
    # ── 결정변수 ──
    P  = opti.variable(3, N+1)                  # base pos (x,y,z)
    Th = opti.variable(3, N+1)                  # base ori (roll,pitch,yaw) 소각
    Ft = {f: opti.variable(3, N+1) for f in FEET}   # 발 위치(world)
    Fr = {f: opti.variable(3, N+1) for f in FEET}   # 접촉력(world)

    g_vec = ca.DM([0,0,-G])
    J = 0
    # ── 초기조건 (시작점 지형 위 BASE_H) ──
    _z0 = BASE_H + float(th(0.0, 0.0))
    opti.subject_to(P[:,0] == ca.DM([0,0,_z0]))
    opti.subject_to(Th[:,0] == ca.DM([0,0,0]))
    for f in FEET:
        opti.subject_to(Ft[f][0,0] == P_NOM[f][0])
        opti.subject_to(Ft[f][1,0] == P_NOM[f][1])
        opti.subject_to(Ft[f][2,0] == th(P_NOM[f][0], P_NOM[f][1]))

    for k in range(N+1):
        pk = P[:,k]; thk = Th[:,k]
        # 총 접촉력·모멘트
        Fsum = ca.DM([0,0,0]); Msum = ca.DM([0,0,0])
        for f in FEET:
            fk = Fr[f][:,k]; rk = Ft[f][:,k]
            Fsum = Fsum + fk
            Msum = Msum + ca.cross(rk - pk, fk)
            if contact[f][k]:
                # 마찰콘(평지 법선 z; 지형법선은 Phase0b): fz>=0, |fxy|<=mu fz
                opti.subject_to(fk[2] >= 0)
                opti.subject_to(fk[2] <= F_MAX)
                opti.subject_to(fk[0] <= MU*fk[2]); opti.subject_to(-fk[0] <= MU*fk[2])
                opti.subject_to(fk[1] <= MU*fk[2]); opti.subject_to(-fk[1] <= MU*fk[2])
                # 지지발: 지형 위 + 정지(발 xy 고정=이전노드와 동일)
                opti.subject_to(rk[2] == th(rk[0], rk[1]))
                if k>0 and contact[f][k-1]:
                    opti.subject_to(Ft[f][0,k] == Ft[f][0,k-1])
                    opti.subject_to(Ft[f][1,k] == Ft[f][1,k-1])
            else:
                opti.subject_to(fk == ca.DM([0,0,0]))      # 스윙: 힘 0
                # 스윙: 지형 위 클리어런스(중간 노드에서 apex)
                opti.subject_to(rk[2] >= th(rk[0], rk[1]) + 0.0)
            # ROM: 발이 base 로컬 공칭±ROM 내 (yaw 소각→월드≈로컬+base_xy)
            dx = rk[0] - (pk[0] + P_NOM[f][0]); dy = rk[1] - (pk[1] + P_NOM[f][1])
            opti.subject_to(dx <=  ROM_DXY[0]); opti.subject_to(-dx <= ROM_DXY[0])
            opti.subject_to(dy <=  ROM_DXY[1]); opti.subject_to(-dy <= ROM_DXY[1])
            opti.subject_to(rk[2]-pk[2] >= FOOT_Z_LO); opti.subject_to(rk[2]-pk[2] <= FOOT_Z_HI)

        # ── SRBD 동역학(중앙차분 가속) ──
        if 0 < k < N:
            acc = (P[:,k+1] - 2*P[:,k] + P[:,k-1])/dt**2
            opti.subject_to(MASS*acc == Fsum + MASS*g_vec)         # 선형
            angacc = (Th[:,k+1] - 2*Th[:,k] + Th[:,k-1])/dt**2
            for i in range(3):
                opti.subject_to(INER[i]*angacc[i] == Msum[i])      # 회전(소각 근사)
        # ── 비용: 전진속도 추종 + 자세/힘 정규화 ──
        if k>0:
            vx = (P[0,k]-P[0,k-1])/dt
            J += 5.0*(vx - vx_des)**2
        J += 50.0*(thk[0]**2 + thk[1]**2 + thk[2]**2)              # roll/pitch/yaw 레벨(강)
        J += 20.0*(P[1,k]**2)                                       # y 드리프트 억제(강)
        # base 높이 = 지지면(플랫폼) 레벨 위 명목(갭에서 안 빠지게 bhref 사용)
        J += 80.0*(P[2,k] - (bhref(P[0,k],P[1,k]) + BASE_H))**2
        # 평활: base 선속도/각속도 최소화(매끄러운 추종 참조)
        if 0<k<N:
            J += 2.0*ca.sumsqr((P[:,k+1]-P[:,k-1])/(2*dt) - ca.DM([vx_des,0,0]))
            J += 5.0*ca.sumsqr((Th[:,k+1]-Th[:,k-1])/(2*dt))
        for f in FEET:
            J += 1e-4*ca.sumsqr(Fr[f][:,k])

    opti.minimize(J)
    # ── 초기추정 ──
    for k in range(N+1):
        xk = x_goal*k/N
        opti.set_initial(P[:,k], [xk, 0, BASE_H + float(bhref(xk,0))])
        for f in FEET:
            fx=xk+P_NOM[f][0]
            fz=float(th(fx,P_NOM[f][1]))
            if _bl is not None and fz < _bl-1e-3: fz=_bl   # 갭 위 init발=플랫폼 레벨(갭바닥 아님)
            opti.set_initial(Ft[f][:,k], [fx, P_NOM[f][1], fz])
            if contact[f][k]:
                opti.set_initial(Fr[f][:,k], [0,0,MASS*G/2])

    opti.solver('ipopt', {'print_time':0}, {'print_level':5 if verbose else 0,
                 'max_iter':800, 'tol':1e-4, 'acceptable_tol':1e-3})
    try:
        sol = opti.solve()
    except Exception as e:
        print("[TOWR] solve 실패:", str(e)[:200])
        try:
            P_=opti.debug.value(P); print("  마지막 x_end=%.3f z범위 %.3f~%.3f"%(P_[0,-1],P_[2].min(),P_[2].max()))
        except Exception: pass
        return None

    out = {'dt':dt,'N':N,'kind':kind,'tkw':tkw,'Tg':Tg,'duty':duty,'gait':gait,
           'P':sol.value(P).tolist(), 'Th':sol.value(Th).tolist(),
           'Ft':{f:sol.value(Ft[f]).tolist() for f in FEET},
           'Fr':{f:sol.value(Fr[f]).tolist() for f in FEET},
           'contact':{f:contact[f] for f in FEET}}
    Pv = sol.value(P)
    print("[TOWR] ✅ solve 성공  x:%.3f→%.3f  z:%.3f~%.3f  총시간%.2fs  vx_avg=%.3f"
          % (Pv[0,0],Pv[0,-1],Pv[2].min(),Pv[2].max(),T,(Pv[0,-1]-Pv[0,0])/T))
    return out


if __name__ == '__main__':
    kind = os.environ.get('TERRAIN','flat')
    gait = os.environ.get('GAIT','trot')
    N    = int(os.environ.get('N','40')); dt=float(os.environ.get('DT','0.02'))
    xg   = float(os.environ.get('XGOAL','0.8'))
    Tg   = float(os.environ.get('TG','0.80' if gait=='crawl' else '0.40'))
    duty = float(os.environ.get('DUTY','0.8' if gait=='crawl' else '0.5'))
    tkw  = {}
    if kind=='step': tkw={'x0':float(os.environ.get('X0','0.6')),'h':float(os.environ.get('H','0.10'))}
    if kind=='gap':  tkw={'x0':float(os.environ.get('X0','0.6')),'x1':float(os.environ.get('X1','0.85')),'depth':float(os.environ.get('DEPTH','0.30'))}
    if kind=='platgap': tkw={'x0':float(os.environ.get('X0','1.0')),'x1':float(os.environ.get('X1','1.3')),'plat':float(os.environ.get('PLAT','0.20'))}
    r = build_and_solve(kind=kind, tkw=tkw, N=N, dt=dt, x_goal=xg, Tg=Tg, duty=duty,
                        gait=gait, verbose=bool(os.environ.get('VERBOSE')))
    if r is not None:
        outf = os.environ.get('OUT','/home/jsh/문서/jsh/simulation/quad/towr/traj_%s.json'%kind)
        with open(outf,'w') as f: json.dump(r,f)
        print("[TOWR] 궤적 저장:", outf)

File: quad/towr/towr_cd.cpp
// TOWR-in-CasADi ★C++ 포팅 — towr_cd.py의 SRBD phase-based 궤적최적화를 CasADi C++로 이식.
//   목적: Python 오버헤드 제거 + C++ 스택 통합. IPOPT/CasADi 그대로. 결과 JSON은 towr_track_B.py 호환.
//   빌드: g++ -O2 towr_cd.cpp -I$ENV/include -L$ENV/lib -lcasadi -o towr_cd  (ENV=proxddp env)
//   실행: TERRAIN=flat GAIT=trot ./towr_cd  (env: TERRAIN GAIT N DT XGOAL TG DUTY X0 X1 H DEPTH PLAT TG_SLOW GAP_R OUT VERBOSE)
#include <casadi/casadi.hpp>
#include <fstream>
#include <sstream>
#include <cmath>
#include <vector>
#include <map>
#include <array>
#include <algorithm>
#include <string>
#include <cstdlib>
#include <cstdio>
using namespace casadi;
using std::string; using std::vector; using std::map;

// ───────── 로봇 SRBD 파라미터 (towr_cd.py와 동일) ─────────
static const double MASS=38.016, G=9.81, BASE_H=0.50, MU=0.6;
static const double INER[3]={0.941,2.521,2.236};
static const double ROM_DXY[2]={0.13,0.09};
static const double FOOT_Z_LO=-0.56, FOOT_Z_HI=-0.40;
static const double F_MAX=2.0*MASS*G;
static const vector<string> FEET={"FL","FR","HL","HR"};
static map<string,std::array<double,2>> P_NOM = {
  {"FL",{{ 0.30, 0.16}}},{"FR",{{ 0.30,-0.16}}},
  {"HL",{{-0.30, 0.16}}},{"HR",{{-0.30,-0.16}}}};

static double envd(const char*k,double d){const char*v=getenv(k);return v?atof(v):d;}
static int    envi(const char*k,int d){const char*v=getenv(k);return v?atoi(v):d;}
static string envs(const char*k,const string&d){const char*v=getenv(k);return v?string(v):d;}

// 지형 높이 h(x) — 심볼릭(MX). y 미사용(평지·terrain=x함수).
static MX terr(const MX& x, const string& kind, const map<string,double>& tk){
  if(kind=="step"){ double x0=tk.at("x0"),h=tk.at("h"),s=40.0; return h*0.5*(1+tanh(s*(x-x0))); }
  if(kind=="gap"){ double x0=tk.at("x0"),x1=tk.at("x1"),d=tk.at("depth"),s=60.0; return -d*0.5*(tanh(s*(x-x0))-tanh(s*(x-x1))); }
  if(kind=="platgap"){ double x0=tk.at("x0"),x1=tk.at("x1"),plat=tk.at("plat"),s=60.0; return plat-plat*0.5*(tanh(s*(x-x0))-tanh(s*(x-x1))); }
  return 0.0*x;   // flat
}
// 지형 높이 — 수치(double)
static double terrd(double x, const string& kind, const map<string,double>& tk){
  if(kind=="step"){ double x0=tk.at("x0"),h=tk.at("h"),s=40.0; return h*0.5*(1+std::tanh(s*(x-x0))); }
  if(kind=="gap"){ double x0=tk.at("x0"),x1=tk.at("x1"),d=tk.at("depth"),s=60.0; return -d*0.5*(std::tanh(s*(x-x0))-std::tanh(s*(x-x1))); }
  if(kind=="platgap"){ double x0=tk.at("x0"),x1=tk.at("x1"),plat=tk.at("plat"),s=60.0; return plat-plat*0.5*(std::tanh(s*(x-x0))-std::tanh(s*(x-x1))); }
  return 0.0;
}

// DM(3×M)을 JSON [[row0],[row1],[row2]] 로 (column-major get_elements: idx=k*3+i)
static string dm2json(const DM& M){
  vector<double> e=M.get_elements(); int nr=M.size1(), nc=M.size2();
  std::ostringstream o; o<<"[";
  for(int i=0;i<nr;i++){ o<<(i?",":"")<<"[";
    for(int k=0;k<nc;k++) o<<(k?",":"")<<e[k*nr+i];
    o<<"]"; }
  o<<"]"; return o.str();
}

int main(){
  string kind=envs("TERRAIN","flat"), gait=envs("GAIT","trot");
  int N=envi("N",40); double dt=envd("DT",0.02);
  double x_goal=envd("XGOAL",0.8);
  double Tg=envd("TG", gait=="crawl"?0.80:0.40);
  double duty=envd("DUTY", gait=="crawl"?0.8:0.5);
  double phase_off=0.0;
  map<string,double> tk;
  if(kind=="step"){ tk["x0"]=envd("X0",0.6); tk["h"]=envd("H",0.10); }
  if(kind=="gap"){ tk["x0"]=envd("X0",0.6); tk["x1"]=envd("X1",0.85); tk["depth"]=envd("DEPTH",0.30); }
  if(kind=="platgap"){ tk["x0"]=envd("X0",1.0); tk["x1"]=envd("X1",1.3); tk["plat"]=envd("PLAT",0.20);
                       tk["tg_slow"]=envd("TG_SLOW",envd("TG",0.40)); tk["gap_r"]=envd("GAP_R",0.45); }

  double T=N*dt, vx_des=x_goal/T;
  bool platgap = (kind=="platgap");
  double _bl = platgap ? tk["plat"] : NAN;
  auto bhref=[&](double x){ return platgap ? _bl : terrd(x,kind,tk); };

  // ── 접촉 스케줄(가변 cadence) ──
  vector<string> CRAWL={"FL","HR","FR","HL"};
  double tgs = tk.count("tg_slow")? tk["tg_slow"] : Tg;
  bool has_gapc = (tk.count("x0") && tgs>Tg);
  double gap_c = has_gapc? 0.5*(tk["x0"]+tk["x1"]) : 0.0;
  double gap_r = tk.count("gap_r")? tk["gap_r"] : 0.45;
  vector<double> ph_arr(N+1,0.0);
  for(int k=1;k<=N;k++){ double xk=x_goal*(k-0.5)/N;
    double tgl=(has_gapc && std::abs(xk-gap_c)<gap_r)? Tg : tgs;
    ph_arr[k]=ph_arr[k-1]+dt/tgl; }
  auto in_stance=[&](const string&foot,int k)->bool{
    double ph=std::fmod(ph_arr[k]+phase_off,1.0); if(ph<0)ph+=1;
    if(gait=="crawl"){ int idx=(int)(std::find(CRAWL.begin(),CRAWL.end(),foot)-CRAWL.begin());
      double win_lo=idx*0.25, sw=0.25*duty; return !(win_lo<=ph && ph<win_lo+sw); }
    double off=(foot=="FL"||foot=="HR")?0.0:0.5;
    double p=std::fmod(ph-off,1.0); if(p<0)p+=1; return p<duty; };
  map<string,vector<int>> contact;
  for(auto&f:FEET){ contact[f].resize(N+1); for(int k=0;k<=N;k++) contact[f][k]=in_stance(f,k)?1:0; }

  // ── Opti 문제 ──
  Opti opti;
  MX P=opti.variable(3,N+1), Th=opti.variable(3,N+1);
  map<string,MX> Ft,Fr;
  for(auto&f:FEET){ Ft[f]=opti.variable(3,N+1); Fr[f]=opti.variable(3,N+1); }
  DM g_vec=DM(vector<double>{0,0,-G});
  MX J=0;

  double z0=BASE_H+terrd(0.0,kind,tk);
  opti.subject_to(P(Slice(),0)==DM(vector<double>{0,0,z0}));
  opti.subject_to(Th(Slice(),0)==DM(vector<double>{0,0,0}));
  for(auto&f:FEET){
    opti.subject_to(Ft[f](0,0)==P_NOM[f][0]);
    opti.subject_to(Ft[f](1,0)==P_NOM[f][1]);
    opti.subject_to(Ft[f](2,0)==terrd(P_NOM[f][0],kind,tk));
  }

  for(int k=0;k<=N;k++){
    MX pk=P(Slice(),k), thk=Th(Slice(),k);
    MX Fsum=DM(vector<double>{0,0,0}), Msum=DM(vector<double>{0,0,0});
    for(auto&f:FEET){
      MX fk=Fr[f](Slice(),k), rk=Ft[f](Slice(),k);
      Fsum=Fsum+fk;
      Msum=Msum+cross(rk-pk,fk);
      if(contact[f][k]){
        opti.subject_to(fk(2)>=0);
        opti.subject_to(fk(2)<=F_MAX);
        opti.subject_to(fk(0)<=MU*fk(2)); opti.subject_to(-fk(0)<=MU*fk(2));
        opti.subject_to(fk(1)<=MU*fk(2)); opti.subject_to(-fk(1)<=MU*fk(2));
        opti.subject_to(rk(2)==terr(rk(0),kind,tk));
        if(k>0 && contact[f][k-1]){
          opti.subject_to(Ft[f](0,k)==Ft[f](0,k-1));
          opti.subject_to(Ft[f](1,k)==Ft[f](1,k-1));
        }
      } else {
        opti.subject_to(fk==DM(vector<double>{0,0,0}));
        opti.subject_to(rk(2)>=terr(rk(0),kind,tk));
      }
      MX dx=rk(0)-(pk(0)+P_NOM[f][0]), dy=rk(1)-(pk(1)+P_NOM[f][1]);
      opti.subject_to(dx<=ROM_DXY[0]); opti.subject_to(-dx<=ROM_DXY[0]);
      opti.subject_to(dy<=ROM_DXY[1]); opti.subject_to(-dy<=ROM_DXY[1]);
      opti.subject_to(rk(2)-pk(2)>=FOOT_Z_LO); opti.subject_to(rk(2)-pk(2)<=FOOT_Z_HI);
    }
    if(0<k && k<N){
      MX acc=(P(Slice(),k+1)-2*P(Slice(),k)+P(Slice(),k-1))/(dt*dt);
      opti.subject_to(MASS*acc==Fsum+MASS*g_vec);
      MX angacc=(Th(Slice(),k+1)-2*Th(Slice(),k)+Th(Slice(),k-1))/(dt*dt);
      for(int i=0;i<3;i++) opti.subject_to(INER[i]*angacc(i)==Msum(i));
    }
    if(k>0){ MX vx=(P(0,k)-P(0,k-1))/dt; J=J+5.0*pow(vx-vx_des,2); }
    J=J+50.0*(pow(thk(0),2)+pow(thk(1),2)+pow(thk(2),2));
    J=J+20.0*pow(P(1,k),2);
    MX bh = platgap? MX(_bl) : terr(P(0,k),kind,tk);
    J=J+80.0*pow(P(2,k)-(bh+BASE_H),2);
    if(0<k && k<N){
      J=J+2.0*sumsqr((P(Slice(),k+1)-P(Slice(),k-1))/(2*dt)-DM(vector<double>{vx_des,0,0}));
      J=J+5.0*sumsqr((Th(Slice(),k+1)-Th(Slice(),k-1))/(2*dt));
    }
    for(auto&f:FEET) J=J+1e-4*sumsqr(Fr[f](Slice(),k));
  }
  opti.minimize(J);

  // ── 초기추정 ──
  for(int k=0;k<=N;k++){
    double xk=x_goal*k/N;
    opti.set_initial(P(Slice(),k), DM(vector<double>{xk,0,BASE_H+bhref(xk)}));
    for(auto&f:FEET){
      double fx=xk+P_NOM[f][0], fz=terrd(fx,kind,tk);
      if(platgap && fz<_bl-1e-3) fz=_bl;
      opti.set_initial(Ft[f](Slice(),k), DM(vector<double>{fx,P_NOM[f][1],fz}));
      if(contact[f][k]) opti.set_initial(Fr[f](Slice(),k), DM(vector<double>{0,0,MASS*G/2}));
    }
  }

  Dict popt, sopt;
  popt["print_time"]=0;
  sopt["print_level"]= getenv("VERBOSE")?5:0;
  sopt["max_iter"]=800; sopt["tol"]=1e-4; sopt["acceptable_tol"]=1e-3;
  opti.solver("ipopt", popt, sopt);

  try{
    OptiSol sol=opti.solve();
    DM Pv=sol.value(P), Thv=sol.value(Th);
    map<string,DM> Ftv,Frv;
    for(auto&f:FEET){ Ftv[f]=sol.value(Ft[f]); Frv[f]=sol.value(Fr[f]); }
    vector<double> pe=Pv.get_elements();
    double x0=pe[0], xe=pe[N*3+0];
    double zmin=1e9,zmax=-1e9; for(int k=0;k<=N;k++){ double z=pe[k*3+2]; zmin=std::min(zmin,z); zmax=std::max(zmax,z);}
    printf("[TOWR-C++] ✅ solve 성공  x:%.3f→%.3f  z:%.3f~%.3f  총시간%.2fs  vx_avg=%.3f\n",
           x0,xe,zmin,zmax,T,(xe-x0)/T);

    // ── JSON 저장 (towr_track_B.py 호환: dt,N,P,Th,Ft,Fr,contact) ──
    string outf=envs("OUT","/home/jsh/문서/jsh/simulation/quad/towr/traj_"+kind+"_cpp.json");
    std::ostringstream o; o.precision(10);
    o<<"{\"dt\":"<<dt<<",\"N\":"<<N<<",\"kind\":\""<<kind<<"\",\"Tg\":"<<Tg<<",\"duty\":"<<duty
     <<",\"gait\":\""<<gait<<"\",\"tkw\":{}";
    o<<",\"P\":"<<dm2json(Pv)<<",\"Th\":"<<dm2json(Thv);
    o<<",\"Ft\":{"; bool first=true; for(auto&f:FEET){ o<<(first?"":",")<<"\""<<f<<"\":"<<dm2json(Ftv[f]); first=false;} o<<"}";
    o<<",\"Fr\":{"; first=true; for(auto&f:FEET){ o<<(first?"":",")<<"\""<<f<<"\":"<<dm2json(Frv[f]); first=false;} o<<"}";
    o<<",\"contact\":{"; first=true; for(auto&f:FEET){ o<<(first?"":",")<<"\""<<f<<"\":[";
      for(int k=0;k<=N;k++) o<<(k?",":"")<<contact[f][k]; o<<"]"; first=false;} o<<"}";
    o<<"}";
    std::ofstream of(outf); of<<o.str(); of.close();
    printf("[TOWR-C++] 궤적 저장: %s\n", outf.c_str());
  }catch(std::exception&e){
    printf("[TOWR-C++] solve 실패: %.200s\n", e.what());
    return 1;
  }
  return 0;
}
